src/v/redpanda/memory_sampling.cc

```diff
@@ -0,0 +1,72 @@
+#include "redpanda/memory_sampling.h"
+
+#include "ssx/future-util.h"
+#include "vlog.h"
+
+#include <seastar/core/future.hh>
+#include <seastar/core/memory.hh>
+#include <seastar/core/sstring.hh>
+#include <seastar/util/memory_diagnostics.hh>
+
+#include <fmt/core.h>
+
+constexpr std::string_view diagnostics_header() {
+    return "Top-N alloc sites - size count stack:\n";
+}
+
+constexpr std::string_view allocation_site_format_str() { return "{} {} {}\n"; }
+
+/// Put `top_n` allocation sites into the front of `allocation_sites`
+void top_n_allocation_sites(
+  std::vector<ss::memory::allocation_site>& allocation_sites, size_t top_n) {
+    std::partial_sort(
+      allocation_sites.begin(),
+      allocation_sites.begin() + top_n,
+      allocation_sites.end(),
+      [](const auto& lhs, const auto& rhs) { return lhs.size > rhs.size; });
+}
+
+ss::sstring memory_sampling::format_allocation_site(
+  const ss::memory::allocation_site& alloc_site) {
+    return fmt::format(
+      allocation_site_format_str(),
+      alloc_site.size,
+      alloc_site.count,
+      alloc_site.backtrace);
+}
+
+memory_sampling::memory_sampling(ss::logger& logger)
+  : _logger(logger) {}
+
+ss::future<> memory_sampling::enable_sampling() {
+    // We chose a sampling rate of ~3MB. From testing this has a very low
+    // overhead of something like ~1%. We could still get away with something
+    // smaller like 1MB and have acceptable overhead (~3%) but 3MB should be a
+    // safer default for the initial rollout.
+    ss::memory::set_heap_profiling_sampling_rate(3000037);
+
+    _timer.set_callback([this]() mutable {
+        auto allocation_sites = ss::memory::sampled_memory_profile();
+
+        const size_t top_n = std::min(size_t(5), allocation_sites.size());
+        top_n_allocation_sites(allocation_sites, top_n);
+
+        ss::sstring output(diagnostics_header());
+
+        for (size_t i = 0; i < top_n; ++i) {
+            output += format_allocation_site(allocation_sites[i]);
+        }
+
+        vlog(_logger.info, "{}", output);
```

Comment:
One thing we could (should) certainly do is keep a bigger buffer around that we reuse across the periodic invocations.

Unfortunately we can't really do a line-by-line buffer thing because we actually want to log the whole line at once as otherwise log lines from multiple shards potentially get intermingled.

Reply:
yeh, good point. @BenPope may have some ideas as our resident fmtlib expert

Reply:
You can [format a range](https://fmt.dev/8.1.1/api.html#ranges-api), there are a few examples of `fmt::join` in the tree. You may do well to implement https://fmt.dev/8.1.1/api.html#formatting-user-defined-types